lib/src/fortio.cpp

```diff
@@ -0,0 +1,337 @@
+
+    ~fguard() noexcept( false ) {
+        if( !this->fp ) return;
+
+        const auto err = std::fsetpos( fp, &this->pos );
+        if( err ) throw std::runtime_error("");
+    }
+};
+
+struct options {
+    bool bigendian  = true;
+    bool size_limit = true;
+    int elemsize    = sizeof( std::int32_t );
+
+    // transform by default unless record is nullptr and this record should be
+    // skipped
+    bool transform    = true;
+    bool force_notail = false;
+    bool allow_notail = false;
+};
+
+options parse_opts( const char* opts ) noexcept {
+    options o;
+
+    for( const char* op = opts; *op != '\0'; ++op ) {
+        switch( *op ) {
+            case 'c': o.elemsize = sizeof( char );         break;
+            case 'i': o.elemsize = sizeof( std::int32_t ); break;
+            case 'f': o.elemsize = sizeof( float );        break;
+            case 'd': o.elemsize = sizeof( double );       break;
+
+            case 'E': o.bigendian = true;                  break;
+            case 'e': o.bigendian = false;                 break;
+
+            case 't': o.transform = true;                  break;
+            case 'T': o.transform = false;                 break;
+
+            case '~': o.force_notail = true;               break;
+            case '$': o.allow_notail = true;               break;
+
+            case '#': o.size_limit = false;                break;
+            default: break;
+        }
+    }
+
+    return o;
+}
+
+}
+
+int eclfio_sizeof( std::FILE* fp, const char* opts, int32_t* out ) try {
+    const auto o = parse_opts( opts );
+
+    fguard guard( fp );
+
+    std::int32_t size;
+    const auto read = std::fread( &size, sizeof( size ), 1, fp );
+    if( read != 1 ) return ECL_ERR_READ;
+
+    if( o.bigendian )
+        size = ntohl( size );
+
+    guard.fp = nullptr;
+
+    *out = size;
+    return ECL_OK;
+
+} catch( std::runtime_error& ) {
+    return ECL_ERR_SEEK;
+}
+
+int eclfio_skip( FILE* fp, const char* opts, int n ) try {
+    fguard guard( fp );
+
+    // TODO: support backwards skips
+    if( n < 0 ) return ECL_EINVAL;
+
+    for( int i = 0; i < n; ++i ) {
+        int err = eclfio_get( fp, opts, nullptr, nullptr );
+        if( err ) return err;
+    }
+
+    guard.fp = nullptr;
+    return ECL_OK;
+} catch( std::runtime_error& ) {
+    return ECL_ERR_SEEK;
+}
+
+int eclfio_get( std::FILE* fp,
+                const char* opts,
+                int32_t* recordsize,
+                void* record ) try {
+
+    auto o = parse_opts( opts );
+
+    // if this is a skip, opt out of transform altogether
+    o.transform = o.transform and bool(record);
+    o.size_limit = o.size_limit and record and recordsize;
+
+    fguard guard( fp );
+
+    std::int32_t head;
+    auto read = std::fread( &head, sizeof( head ), 1, fp );
+    if( read != 1 ) return ECL_ERR_READ;
+
+    if( o.bigendian )
+        head = (std::int32_t)ntohl( (std::uint32_t)head );
+
+    if( head < 0 ) return ECL_INVALID_RECORD;
```

Comment:
This needs to check for int32 overflow too.

Reply:
> This needs to check for int32 overflow too.

OK; is the head and tail unsigned in the Fortran standard?